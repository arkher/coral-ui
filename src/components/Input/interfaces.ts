import { BoxProps } from '@shopify/restyle';
import { TextInputProps, TextInput } from 'react-native';
import { Theme } from '../../themes/default';

type InputProps = {
  variant: CustomHeightComponent;
} & Partial<{
  placeholder: string;
  icon: string;
  bw: CustomBorderWidth;
  p: CustomSpacing;
}> &
  Partial<BoxProps<Theme>> &
  TextInputProps;

type InputRef = HTMLInputElement &
  TextInput & {
    focus(): void;
    blur(): void;
    error(): void;
    success(): void;
    clearStatus(): void;
  };

export { InputProps, InputRef };
