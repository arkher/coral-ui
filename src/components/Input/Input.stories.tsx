import React from 'react';
import { withDesign } from 'storybook-addon-designs';
import {
  COLORS,
  BORDER_WIDTH,
  SPACINGS,
  RADIUS,
  HEIGHTS,
} from '../../themes/tokens';
import Input from './Input';

export default {
  title: 'Elements/Input',
  component: Input,
  argTypes: {
    placeholder: {
      control: { type: 'text' },
    },
    borderColor: {
      options: COLORS,
      control: { type: 'select' },
    },
    borderWidth: {
      options: BORDER_WIDTH,
      control: { type: 'select' },
    },
    padding: {
      options: SPACINGS,
      control: { type: 'select' },
    },
    borderRadius: {
      options: RADIUS,
      control: { type: 'select' },
    },
    variant: {
      options: HEIGHTS,
      control: { type: 'select' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    multiline: {
      control: { type: 'boolean' },
    },
    numberOfLines: {
      control: { type: 'number' },
    },
  },
  decorators: [withDesign],
};

interface TextInputProps {
  placeholder: string;
  value: string;
  borderColor: Custom.Colors;
  borderWidth: Custom.BorderWidth;
  padding: 'nano' | 'sm' | 'md' | 'lg' | 'xs' | 'quark';
  borderRadius: Custom.Radius;
  variant: Custom.HeightComponent;
  disabled: boolean;
  multiline: boolean;
  numberOfLines: number;
}

export const Default = ({
  placeholder,
  value,
  borderColor,
  borderWidth,
  padding,
  borderRadius,
  variant,
  disabled,
  multiline,
  numberOfLines,
}: TextInputProps): React.ReactNode => (
  <Input
    placeholder={placeholder || 'Placeholder'}
    borderColor={borderColor}
    value={value}
    bw={borderWidth || 'hairline'}
    p={padding || 'nano'}
    borderRadius={borderRadius}
    h={variant}
    editable={disabled}
    multiline={multiline}
    numberOfLines={numberOfLines}
  />
);
