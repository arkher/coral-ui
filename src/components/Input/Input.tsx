import React, {
  useCallback,
  useRef,
  useState,
  useImperativeHandle,
  forwardRef,
  ForwardRefRenderFunction,
} from 'react';

import { createRestyleComponent, spacing, useTheme } from '@shopify/restyle';
import { Keyboard, TextInput, TouchableWithoutFeedback } from 'react-native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';
import { borderWidth, Theme } from '../../themes/default';
import Box from '../Box';
import { InputFowardEvents, InputProps, InputRef } from './interfaces';

const Input: React.FC<InputProps> = (
  {
    placeholder,
    variant,
    editable,
    multiline,
    numberOfLines,
    icon,
    maxLength,
    keyboardType,
    autoCapitalize,
    style,
    returnKeyType,
    ...props
  },
  ref,
) => {
  const { colors } = useTheme<Theme>();
  const [isFocused, setIsFocused] = useState(false);
  const [isFilled, setIsFilled] = useState(false);

  const [hasError, setHasError] = useState(false);
  const [hasSuccess, setHasSuccess] = useState(false);

  const [colorStatus, setColorStatus] =
    useState<keyof Theme['colors']>('neutralDark');

  const inputElementRef = useRef<InputRef>(null);

  const handleInputFocus = useCallback(() => {
    setIsFocused(true);

    if (!hasError && !hasSuccess) {
      setColorStatus('primaryBase');
    }
  }, [hasError, hasSuccess]);

  const handleInputBlur = useCallback(() => {
    setIsFocused(false);

    if (!hasError && !hasSuccess) {
      setColorStatus('neutralDark');
    }
  }, [hasError, hasSuccess]);

  const handleClear = useCallback(() => {
    inputElementRef.current?.clear();
    setIsFilled(false);
  }, []);

  const handleChange = useCallback(
    (newValue: string) => {
      if (maxLength && newValue?.length <= maxLength) {
        // eslint-disable-next-line no-param-reassign
        ref.current.value = newValue;
      }

      if (newValue.length > 0) {
        setIsFilled(true);
      }
    },
    [maxLength, ref],
  );

  useImperativeHandle(
    ref,
    () =>
      ({
        focus: () => {
          handleInputFocus();
        },
        blur: () => {
          handleInputBlur();
        },
        error: () => {
          setHasError(true);
          setColorStatus('feedbackErrorBase');
        },
        success: () => {
          setHasSuccess(true);
          setColorStatus('feedbackSuccessBase');
        },
        clear: () => {
          setHasError(false);
          setHasSuccess(false);
          setColorStatus('neutralDark');
        },
      } as InputFowardEvents),
  );

  return (
    <Box
      bw={hasError || hasSuccess || isFocused ? 'thin' : 'hairline'}
      borderColor={colorStatus}
      borderRadius="sm"
      flexDirection="row"
      alignItems="center"
      h={variant}
      {...props}
    >
      <TextInput
        testID="Input"
        ref={inputElementRef}
        placeholder={placeholder}
        placeholderTextColor={colors.neutralDark}
        onBlur={handleInputBlur}
        onSubmitEditing={() => {
          Keyboard.dismiss();
        }}
        onFocus={handleInputFocus}
        onChangeText={handleChange}
        editable={editable}
        multiline={multiline}
        maxLength={maxLength}
        keyboardType={keyboardType}
        returnKeyType={returnKeyType}
        numberOfLines={numberOfLines}
        autoCapitalize={autoCapitalize}
        selectionColor={colors.neutralDark}
        style={style}
      />

      {icon && (
        <TouchableWithoutFeedback onPress={handleClear}>
          <Box ml="quarck">
            <Icon
              name={icon}
              size={24}
              color={
                isFocused || isFilled ? colors.primaryBase : colors.neutralDark
              }
            />
          </Box>
        </TouchableWithoutFeedback>
      )}
    </Box>
  );
};

export default createRestyleComponent<InputProps, Theme>(
  [spacing, borderWidth],
  forwardRef(Input as ForwardRefRenderFunction<InputProps>),
);
