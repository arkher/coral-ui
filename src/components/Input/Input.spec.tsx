import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react-native';
import { ThemeProvider } from '@shopify/restyle';
import Input from './Input';
import themeMaestro from '../../themes/maestro';

afterEach(cleanup);

jest.mock('react-native-vector-icons/MaterialCommunityIcons', () => 'Icon');

test('should have a placeholder passed by prop', () => {
  const { getAllByPlaceholderText } = render(
    <ThemeProvider theme={themeMaestro}>
      <Input placeholder="Placeholder" />
    </ThemeProvider>,
  );

  expect(getAllByPlaceholderText('Placeholder').length).toBe(1);
});

test('should render input from user', () => {
  const CHANGE_TEXT = 'Texto inserido pelo usu√°rio';
  let value = 'Texto inicial';

  const { getByTestId } = render(
    <ThemeProvider theme={themeMaestro}>
      <Input
        placeholder="Placeholder"
        value={value}
        onChangeText={newValue => {
          value = newValue;
        }}
      />
    </ThemeProvider>,
  );

  expect(getByTestId('Input').props.value).toBe(value);

  fireEvent.changeText(getByTestId('Input'), CHANGE_TEXT);
  expect(value).toBe(CHANGE_TEXT);
});

test('should Input render correctly', async () => {
  const { toJSON } = render(
    <ThemeProvider theme={themeMaestro}>
      <Input placeholder="Placeholder" />
    </ThemeProvider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
