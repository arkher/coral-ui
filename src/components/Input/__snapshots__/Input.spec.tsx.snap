// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Input render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "borderColor": "#5D6A75",
        "borderRadius": 4,
        "display": "flex",
        "flexDirection": "row",
      },
      Array [
        Object {
          "borderWidth": 1,
          "height": 56,
        },
      ],
    ]
  }
  testID="Box"
>
  <TextInput
    allowFontScaling={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    onSubmitEditing={[Function]}
    placeholder="Placeholder"
    placeholderTextColor="#5D6A75"
    rejectResponderTermination={true}
    selectionColor="#5D6A75"
    style={
      Array [
        Object {},
      ]
    }
    testID="Input"
    underlineColorAndroid="transparent"
  />
</View>
`;
